### PR TITLE
Add support for C++20 Compilation

### PR DESCRIPTION
**Pull request by QuantamHD**
Couple of notes on C++20

* LEMON 1.3.1 has an issue that needs to be resolved related to allocators. (Included a patch from https://lemon.cs.elte.hu/trac/lemon/ticket/631#no1)
* fmt in versions greater than 8 added the ability to constexpr compile format lines. The downside is runtime strings need to be wrapped in fmt::runtime. Added FMT_RUNTIME macro to support both versions. The current version of spd_log is on version 7 of FMT
* template ids on constructors and destructors have been removed. Removed usage in abk libs.

Introduced a github action to build openroad on c++20. Should prevent any usage of non-supported c++20 features, but this does not mean the project is moving to c++20 yet.

**OpenROAD will stay on C++17**

### CODE REVIEW

**Comment by QuantamHD** [comment]
ABK required a format see https://github.com/The-OpenROAD-Project/OpenROAD/pull/2785/commits/196ce1dad35a6985e9a23ea469f844b68db8675b#diff-f5e647735a7b1efd76bcb1f5f3a12f573e699220fa7bfbc3e68a1e9f339aa6d1

for C++20 changes only

**Comment by maliberty** [comment]
@vvbandeira please comment on the action & build changes.

File: third-party/lemon/patches/allocator-patch.patch
Comment (maliberty) [inline third-party/lemon/patches/allocator-patch.patch:1]: Is there a new lemon release we should upgrade to rather than apply one patch?

Comment (QuantamHD) [reply third-party/lemon/patches/allocator-patch.patch:1]: The last lemon release was in 2014. There are commits after this point, but this patch was never merged in.

File: src/utl/include/utl/Logger.h
Comment (maliberty) [inline src/utl/include/utl/Logger.h:96]: Kinda icky but hopefully the compile time checking is beneficial.

File: third-party/lemon/patches/allocator-patch.patch
Comment (QuantamHD) [reply third-party/lemon/patches/allocator-patch.patch:1]: Also important to note that this patch is only applied in the action not the DependencyInstaller script which is intentional.

Really my goal was to get the CI action up and running and not to support c++20 in the project directly.

File: src/utl/include/utl/Logger.h
Comment (QuantamHD) [reply src/utl/include/utl/Logger.h:96]: You only need to add this macro when you are passing a non compile const string, or maybe said another way a dynamic template.

File: third-party/lemon/patches/allocator-patch.patch
Comment (gadfort) [reply third-party/lemon/patches/allocator-patch.patch:1]: Does it make sense to fork that project and apply the patch there instead? It doesn't seem like they are maintaining that project and if it's a required dependency we need the ability to fix issues like this in a better way than a patch getting applied during an install step

File: src/utl/include/utl/Logger.h
Comment (gadfort) [reply src/utl/include/utl/Logger.h:96]: Can we add some developer documentation about this requirement in utl? Or could we provide a static function in utl to do this automatically? I worry developer not on c++20 will not know this and break things for you.

Comment (gadfort) [reply src/utl/include/utl/Logger.h:96]: Can we add some developer documentation about this requirement in utl? Or could we provide a static function in utl to do this automatically? I worry developer not on c++20 will not know this and break things for you.

Comment (gadfort) [reply src/utl/include/utl/Logger.h:96]: Can we add some developer documentation about this requirement in utl? Or could we provide a static function in utl to do this automatically? I worry developer not on c++20 will not know this and break things for you.

Comment (maliberty) [reply src/utl/include/utl/Logger.h:96]: https://github.com/The-OpenROAD-Project/OpenROAD/blob/master/docs/contrib/Logger.md would be a good place.  I think the added CI is to prevent such mistakes.

File: third-party/lemon/patches/allocator-patch.patch
Comment (maliberty) [reply third-party/lemon/patches/allocator-patch.patch:1]: That sounds reasonable.  I'm sure there will be other users of c++20 - eventually even the project itself.

Comment (QuantamHD) [reply third-party/lemon/patches/allocator-patch.patch:1]: Are we suggesting dumping the files in this repo, or adding a fork?

Comment (maliberty) [reply third-party/lemon/patches/allocator-patch.patch:1]: I would make a new repo in GH.

Comment (QuantamHD) [reply third-party/lemon/patches/allocator-patch.patch:1]: I'm a little bit torn on this particular issues the LEMON mercurial repo does have recent commits, but no new releases. 

I'm going to open a separate issue where we can decide what we want to do.

http://lemon.cs.elte.hu/hg/lemon

Comment (gadfort) [reply third-party/lemon/patches/allocator-patch.patch:1]: I guess my point is this patch was proposed to them 3 years and it's still not in any release. I see the commits, but they don't appear to be that responsive. Please reference the issue when created here.

File: .github/workflows/github-actions-c++.yml
Comment (maliberty) [reply .github/workflows/github-actions-c++.yml:5]: I suppose that time is mostly in setting up the dependencies.  It would be better to have a prepared container to skip that overhead per build.

Comment (QuantamHD) [reply .github/workflows/github-actions-c++.yml:5]: 5 minutes or so is in the dependencies. Which a container would help resolve. I could also look into switching the build from make to ninja.

Comment (maliberty) [reply .github/workflows/github-actions-c++.yml:5]: It should not take 40min to compile OR.  For comparison it is ~10min total to build in docker for our platforms in Jenkins.  The native builds at 5 minutes.

File: third-party/lemon/patches/allocator-patch.patch
Comment (QuantamHD) [reply third-party/lemon/patches/allocator-patch.patch:1]: https://github.com/The-OpenROAD-Project/OpenROAD/issues/2791

File: .github/workflows/github-actions-c++.yml
Comment (QuantamHD) [reply .github/workflows/github-actions-c++.yml:5]: How many cores do those machines have? I think the Github runners are limited to 2.

Comment (QuantamHD) [reply .github/workflows/github-actions-c++.yml:5]: Got the compile time down to 28 minutes.

File: src/utl/include/utl/Logger.h
Comment (QuantamHD) [reply src/utl/include/utl/Logger.h:96]: Done.

File: etc/Build.sh
Comment (QuantamHD) [reply etc/Build.sh]: Done.

File: third-party/lemon/patches/allocator-patch.patch
Comment (QuantamHD) [reply third-party/lemon/patches/allocator-patch.patch:1]: Discussion happening in https://github.com/The-OpenROAD-Project/OpenROAD/issues/2791

**Comment by QuantamHD** [comment]
@vvbandeira I updated the spdlog version in the Dependency Installer script to v1.10.0. @maliberty signed off.

Letting you know so you can update the docker images in Jenkins.

**Comment by maliberty** [comment]
@vvbandeira we'll need to update the CI images for the spdlog update

**Comment by QuantamHD** [comment]
@maliberty This is ready to go on my side let me know if you need any more changes.

**Comment by maliberty** [comment]
lgtm but I'll leave it to @vvbandeira to approve and coordinate the spdlog upgrade.

**Comment by cdleary** [comment]
Just noting I was looking to use `std::span` which I believe is available as of C++20, in case the motivation is helpful. :-)

**Comment by QuantamHD** [comment]
@vvbandeira Any objections to merging?

**Comment by QuantamHD** [comment]
@cdleary std::span (gcc10) appears to not be supported by gcc8 on centos7. Which we need to continue to support. `boost::span` appears to be the next best option.

**Comment by maliberty** [comment]
@cdleary we are only saying you can compile OR with c++, not that we are moving to c++20.  Any c++20 or later features should not be used in OR.

File: docs/contrib/Logger.md
Comment (rovinski) [inline docs/contrib/Logger.md:42]:
```suggestion
OpenROAD uses `spdlog` which uses `fmt_lib` under the hood. Below is an example of
what is no longer allowed.
```

File: docs/contrib/Logger.md
Comment (rovinski) [inline docs/contrib/Logger.md:44]:
```suggestion
In order to make use of runtime format strings, we have introduced a 
```

File: docs/contrib/Logger.md
Comment (rovinski) [inline docs/contrib/Logger.md:46]:
```suggestion
`FMT_RUNTIME` macro in Logger.h. You should use this macro any time you
pass a dynamic string as the format string 
```

**Comment by QuantamHD** [comment]
@vvbandeira gentle ping on CI update for the C++20 patch. I think it's ready to go.

**Comment by vvbandeira** [comment]
@QuantamHD, apologies for the delay. I have started to generate the new CI image, but since I am travelling today, it will be best to do the switch tomorrow when I have a more stable connection to address any issues.

**Comment by QuantamHD** [comment]
No rush just wanted to make sure it was on your radar. Thanks for taking a look

**Review by vvbandeira — APPROVED** [review]

**Pull request merged**
